certora/specs/Reverts.spec

```diff
@@ -0,0 +1,313 @@
+           newFeeRecipient == oldFeeRecipient ||
+           (fee != 0 && newFeeRecipient == 0)
+        => lastReverted;
+}
+
+rule submitGuardianRevertCondition(env e, address newGuardian) {
+    address owner = owner();
+    address oldGuardian = guardian();
+    uint64 pendingGuardianValidAt;
+    _, pendingGuardianValidAt = pendingGuardian();
+
+    // Safe require because it is a verified invariant.
+    require isTimelockInRange();
+    // Safe require as it corresponds to year 2262.
+    require e.block.timestamp < 2^63;
+
+    submitGuardian@withrevert(e, newGuardian);
+
+    assert lastReverted <=>
+        e.msg.value != 0 ||
+        e.msg.sender != owner ||
+        newGuardian == oldGuardian ||
+        pendingGuardianValidAt != 0;
+}
+
+rule submitCapInputValidation(env e, MetaMorphoHarness.MarketParams marketParams, uint256 newSupplyCap) {
+    MorphoHarness.Id id = Morpho.libId(marketParams);
+
+    bool hasCuratorRole = hasCuratorRole(e.msg.sender);
+    address asset = asset();
+    uint256 lastUpdate = Morpho.lastUpdate(id);
+    uint256 pendingCapValidAt;
+    _, pendingCapValidAt = pendingCap(id);
+    uint184 supplyCap;
+    uint64 removableAt;
+    supplyCap, _, removableAt = config(id);
+
+    submitCap@withrevert(e, marketParams, newSupplyCap);
+
+    assert e.msg.value != 0 ||
+           !hasCuratorRole ||
+           marketParams.loanToken != asset ||
+           lastUpdate == 0 ||
+           pendingCapValidAt != 0 ||
+           removableAt != 0 ||
+           newSupplyCap == assert_uint256(supplyCap)
+        => lastReverted;
+}
+
+rule submitMarketRemovalRevertCondition(env e, MetaMorphoHarness.MarketParams marketParams) {
+    MorphoHarness.Id id = Morpho.libId(marketParams);
+
+    bool hasCuratorRole = hasCuratorRole(e.msg.sender);
+    uint256 pendingCapValidAt;
+    _, pendingCapValidAt = pendingCap(id);
+    uint184 supplyCap;
+    bool enabled;
+    uint64 oldRemovableAt;
+    supplyCap, enabled, oldRemovableAt = config(id);
+
+    // Safe require because it is a verified invariant.
+    require isTimelockInRange();
+    // Safe require as it corresponds to year 2262.
+    require e.block.timestamp < 2^63;
+
+    submitMarketRemoval@withrevert(e, marketParams);
+
+    assert lastReverted <=>
+        e.msg.value != 0 ||
+        !hasCuratorRole ||
+        pendingCapValidAt != 0 ||
+        supplyCap != 0 ||
+        !enabled ||
+        oldRemovableAt != 0;
+}
+
+rule setSupplyQueueInputValidation(env e, MorphoHarness.Id[] newSupplyQueue) {
+    bool hasAllocatorRole = hasAllocatorRole(e.msg.sender);
+    uint256 maxQueueLength = maxQueueLength();
+    uint256 i;
+    require i < newSupplyQueue.length;
+    uint184 anyCap;
+    anyCap, _, _ = config(newSupplyQueue[i]);
+
+    setSupplyQueue@withrevert(e, newSupplyQueue);
+
+    assert e.msg.value != 0 ||
+           !hasAllocatorRole ||
+           newSupplyQueue.length > maxQueueLength ||
+           anyCap == 0
+        => lastReverted;
```

Comment:
why could it fail otherwise?

Reply:
I added a comment: it cannot revert for another reason, but we cannot show that because we verify `forall i. cap[i] = 0 => revert` and not `(forall i. cap[i] = 0) => revert`. The formula `forall i. revert => cap[i] = 0` is false, or, said otherwise: it can revert because of any index, but we cannot know the index in advance.